lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -441,6 +1975,244 @@ bool StackAnalysis::runOnModule(Module &M) {
+  UnexpectedPCMarker->addFnAttr(llvm::Attribute::ReadOnly);
+  UnexpectedPCMarker->addFnAttr(llvm::Attribute::NoUnwind);
+
+  // Queue to be populated with the CFEP
+  llvm::SmallVector<BasicBlock *, 8> Worklist;
+  llvm::SmallVector<BasicBlockNode *, 4> EntryBlocks;
+  UniquedQueue<BasicBlockNode *> CFEPQueue;
+  SmallCallGraph CG{};
+
+  using BasicBlockToNodeMapTy = llvm::DenseMap<BasicBlock *, BasicBlockNode *>;
+  BasicBlockToNodeMapTy BasicBlockNodeMap;
+
+  // Create over-approximated call graph
+  for (CFEP &Function : Functions) {
+    BasicBlockNode Node{ Function.Entry };
+    auto *GraphNode = CG.addNode(Node);
+    BasicBlockNodeMap[Function.Entry] = GraphNode;
+  }
+
+  for (CFEP &Function : Functions) {
+    llvm::SmallSet<BasicBlock *, 8> Visited;
+    auto *StartNode = BasicBlockNodeMap[Function.Entry];
+    Worklist.emplace_back(Function.Entry);
+
+    do {
+      auto *Current = Worklist.pop_back_val();
+      Visited.insert(Current);
+      for (auto *Successor : successors(Current)) {
+        if (isFunctionCall(Current)) {
+          auto *Callee = getFunctionCallCallee(Current);
+          // If not an indirect call, add the node to the CG
+          if (Callee) {
+            auto *Node = BasicBlockNodeMap[Callee];
+            StartNode->addSuccessor(Node);
+          }
+          auto *Next = getFallthrough(Current);
+          if (!Visited.count(Next))
+            Worklist.push_back(Next);
+          continue;
+        }
+
+        if (!GCBI::isPartOfRootDispatcher(Successor)
+            && !Visited.count(Successor))
+          Worklist.push_back(Successor);
+      }
+    } while (!Worklist.empty());
+  }
+
+  for (const auto &[_, Node] : BasicBlockNodeMap) {
+    if (!Node->hasPredecessors())
+      EntryBlocks.emplace_back(Node);
+
+    // Recursive?
+    if (*Node->predecessors().begin() == Node)
+      EntryBlocks.emplace_back(Node);
+  }
+
+  BasicBlockNode FakeNode{ nullptr };
+  auto *RootNode = CG.addNode(FakeNode);
+  CG.setEntryNode(RootNode);
+  for (const auto &Node : EntryBlocks)
+    RootNode->addSuccessor(Node);
+
+  if (StackAnalysisLog.isEnabled()) {
+    std::error_code EC;
+    raw_fd_ostream Out(M.getModuleIdentifier() + "_CG.dot",
+                       EC,
+                       llvm::sys::fs::OF_None);
+    revng_assert(!EC);
+    llvm::WriteGraph(Out, &CG);
+    revng_log(StackAnalysisLog, "Call graph written on disk");
+    Out.close();
+  }
+
+  for (auto *Node : llvm::post_order(&CG)) {
+    if (Node->hasPredecessors())
+      CFEPQueue.insert(Node);
+  }
+
+  // Collect all the ABI registers
+  std::vector<llvm::GlobalVariable *> ABIRegisters;
+  for (auto *CSV : GCBI.abiRegisters())
+    if (CSV && !(GCBI.isSPReg(CSV)))
+      ABIRegisters.emplace_back(CSV);
+
+  std::optional<int64_t> FSO;
+  switch (GCBI.arch()) {
+  case llvm::Triple::x86: {
+    FSO = 4;
+    break;
+  }
+  case llvm::Triple::x86_64: {
+    FSO = 8;
+    break;
+  }
+  default: {
+    FSO = 0;
+    break;
+  }
+  }
```

Comment:
Concentrate architecture specific things in a couple of functions in a well defined place with an appropriate `TODO`.

Reply:
Can you already tell where this should go? `BinaryFile` and again a new field in `revng.input.architecture` or somewhere else?

Reply:
OK, that's even better. You can just redo the thing you did starting from `BinaryFiles.{cpp.h}` down to here.